### PR TITLE
Feature/um 716 improve destruction resilliance of primary pools

### PR DESCRIPTION
**Pull request by mcfadden8**
1. Added logging to primary pools to better reflect pool life cycle
2. Added (missing) REPLAY of coalesce operation to `QuickPool`
3. Modified destructors of primary pools to try/catch/log/ignore deallocation failures from the device which may happen if destruction occurrs after the device's runtime has already shut down.

### CODE REVIEW

**Comment by mcfadden8** [comment]
LGTM

File: src/umpire/strategy/DynamicPoolMap.cpp [hunk 1]
```diff
@@ -35,10 +35,20 @@ DynamicPoolMap::DynamicPoolMap(
       m_next_minimum_pool_allocation_size{
           aligned_round_up(next_minimum_pool_allocation_size)}
 {
+  UMPIRE_LOG(Debug, " { "
```
Comment (davidbeckingsale) [inline src/umpire/strategy/DynamicPoolMap.cpp:38]: Can you reformat this to look like how the debug constructor log normally looks? I think the format you used before is arg=val for all the arguments.

File: src/umpire/strategy/DynamicPoolMap.cpp [hunk 2]
```diff
@@ -301,6 +234,129 @@ MemoryResourceTraits DynamicPoolMap::getTraits() const noexcept
...
+  try {
+    aligned_deallocate(ptr);
+  }
+  catch (...) {
+    if (m_is_destructing) {
+      //
+      // Ignore the error if we are destructing as the resource to which we
+      // are releasing to may have gone away
+      //
+      UMPIRE_LOG(Error, "Pool is destructing, Exception Ignored");
```
Comment (davidbeckingsale) [inline src/umpire/strategy/DynamicPoolMap.cpp:352]: I think you should print the exception message. The underlying allocator will (and should) always exist. The comment above doesn't reflect the real issue this is fixing.

File: src/umpire/strategy/DynamicSizePool.hpp [hunk 3]
```diff
@@ -226,10 +216,27 @@ class DynamicSizePool : private umpire::strategy::mixins::AlignedAllocation {
...
+        try {
+          aligned_deallocate(curr->data);
+        }
+        catch (...) {
+          if (m_is_destructing) {
+            //
+            // Ignore the error if we are destructing as the resource to which we
+            // are releasing to may have gone away
+            //
+            UMPIRE_LOG(Error, "Pool is destructing, Exception Ignored");
```
Comment (davidbeckingsale) [inline src/umpire/strategy/DynamicSizePool.hpp:234]: Same comment as other pool.

File: src/umpire/strategy/DynamicSizePool.hpp [hunk 4]
```diff
@@ -282,44 +289,53 @@ class DynamicSizePool : private umpire::strategy::mixins::AlignedAllocation {
         m_next_minimum_pool_allocation_size{
             aligned_round_up(next_minimum_pool_allocation_size)}
   {
+    UMPIRE_LOG(Debug, " { "
```
Comment (davidbeckingsale) [inline src/umpire/strategy/DynamicSizePool.hpp:292]: Reformat message please.

File: src/umpire/strategy/QuickPool.cpp [hunk 5]
```diff
@@ -28,16 +27,26 @@ QuickPool::QuickPool(const std::string& name, int id, Allocator allocator,
       m_next_minimum_pool_allocation_size{
           aligned_round_up(next_minimum_pool_allocation_size)}
 {
+  UMPIRE_LOG(Debug, " { "
```
Comment (davidbeckingsale) [inline src/umpire/strategy/QuickPool.cpp:30]: Reformat.

File: src/umpire/strategy/QuickPool.cpp [hunk 6]
```diff
@@ -207,7 +216,22 @@ void QuickPool::release()
...
+      try {
+        aligned_deallocate(chunk->data);
+      }
+      catch (...) {
+        if (m_is_destructing) {
+          //
+          // Ignore the error if we are destructing as the resource to which we
+          // are releasing to may have gone away
+          //
+          UMPIRE_LOG(Error, "Pool is destructing, Exception Ignored");
```
Comment (davidbeckingsale) [inline src/umpire/strategy/QuickPool.cpp:229]: Fix comment, log the exception message.

**Review by davidbeckingsale — CHANGES_REQUESTED** [review]

File: src/umpire/strategy/DynamicPoolMap.cpp (hunk 2 above)
Comment (mcfadden8) [reply src/umpire/strategy/DynamicPoolMap.cpp:352]: I will fix the comment, but the underlying umpire exception is already printed right (by the call to UMPIRE_ERROR)

Comment (davidbeckingsale) [reply src/umpire/strategy/DynamicPoolMap.cpp:352]: Yup, you're correct.

Comment (mcfadden8) [reply src/umpire/strategy/DynamicPoolMap.cpp:352]: and sorry, but the comment is correct.  I will change it to: "...the resource to which we are releasing from...".  The 'resource' could be anything that offers a mechanism external to Umpire to cause it to be reset.

File: src/umpire/strategy/QuickPool.cpp (hunk 6 above)
Comment (mcfadden8) [reply src/umpire/strategy/QuickPool.cpp:229]: Are you asking me to distinguish between an Umpire exception and some other type of exception?  If it is an Umpire exception, the message will be logged by UMPIRE_ERROR.

I don't understand what you are looking for in the comment.  Please suggest an alternative?

Comment (mcfadden8) [reply src/umpire/strategy/QuickPool.cpp:230]: Done

File: src/umpire/strategy/QuickPool.cpp (hunk 5 above)
Comment (mcfadden8) [reply src/umpire/strategy/QuickPool.cpp]: Done

File: src/umpire/strategy/DynamicSizePool.hpp (hunk 4 above)
Comment (mcfadden8) [reply src/umpire/strategy/DynamicSizePool.hpp]: Done

File: src/umpire/strategy/DynamicSizePool.hpp (hunk 3 above)
Comment (mcfadden8) [reply src/umpire/strategy/DynamicSizePool.hpp:234]: Done

File: src/umpire/strategy/DynamicPoolMap.cpp (hunk 2 above)
Comment (mcfadden8) [reply src/umpire/strategy/DynamicPoolMap.cpp:354]: Done

File: src/umpire/strategy/DynamicPoolMap.cpp (hunk 1 above)
Comment (mcfadden8) [reply src/umpire/strategy/DynamicPoolMap.cpp]: Done

**Comment by mcfadden8** [comment]
LGTM

**Review by kab163 — APPROVED** [review]
Looks good!

**Review by davidbeckingsale — APPROVED** [review]

**Pull request merged**